// departures-board/departures_board.js
class departures_board{
	constructor(num_characters, is_animated, timing, auto_refresh, auto_refresh_period, dark_tiles, size, force_all_caps){
		this.num_characters = num_characters;
		this.is_animated = is_animated;
		this.auto_refresh = auto_refresh;
		this.auto_refresh_period = auto_refresh_period;
		this.timing = timing;
		this.size = size;
		this.dark_tiles = dark_tiles;
		this.force_all_caps = force_all_caps;
		this.id = this.createUUID();
		this.caption="";
	}
	
	setText(data){
			var SplunkVisUtils = require('api/SplunkVisualizationUtils');
			var vizUtils = require('api/SplunkVisualizationUtils');
			this.caption = ""
			
			try{
				//------------------------------  Get data row field indexes ----------------------------------------------------------------------
				this.caption = data.rows[0][0];
				if(this.force_all_caps){
					this.caption = this.caption.toUpperCase();
				}
			} catch(err) {
				console.log("Error setting data. " + err);
			}
		}
	
	getHTML(){
		var styles = (this.dark_tiles) ? "header dark " + this.size : "header light " + this.size;
		var html = "<input class=\"" + styles +"\" id=\"" + this.id +"\" />";
		return html;
		
	}
	
	createUUID() {
		var s = [];
		var hexDigits = "0123456789abcdef";
		for (var i = 0; i < 10; i++) {
			s[i] = hexDigits.substr(Math.floor(Math.random() * 0x10), 1);
		}
		var uuid = s.join("");
		return "db_" + uuid;
	}
	
	start(){
		//const flap = require('flapper');
		var caption = this.caption;
		var id = this.id;
		var auto_refresh_period = this.auto_refresh_period;
		window.jQuery("#" + this.id).val(caption)
		window.jQuery("#" + this.id).flapper({
			width: this.size,
			chars_preset: 'alphanum',
			transform: this.is_animated,
			timing: this.timing
		});
		if(this.auto_refresh) {
			setTimeout(function(){
				setInterval(function(){
					window.jQuery("#" + this.id).val('').change();						
					setTimeout(function(){window.jQuery("#" + this.id).val(caption).change();},1000);
				}, auto_refresh_period * 1000);
			}, 1000);
		} // End if
	}
		
	
	
}
/*
$(document).ready(function() {
                var $el = $('#' + this.id);
                $el.flapper({
                    width: this.size,
                    chars_preset: 'alphanum',
					transform: this.animate,
					timing: this.timing
                });

                setTimeout(function(){
                    $el.val('FLAPPER').change();
                    var toggle = false;
                    setInterval(function(){
                        if (toggle) {
                            $el.val(this.caption).change();
                        } else {
                            $el.val('X' + this.caption).change();
                        }
                        toggle = !toggle;
                    }, 5000);
                }, 1000);
            });
			*/

	module.exports =  departures_board;
